Update and pin to Ubuntu Jammy

yvanzo opened the pull request:
## Problem

The repository was still functional but outdated (Ubuntu Xenial) and missing some basics for proper maintenance.

## Solution

* Pin base image tag
* Use up-to-date packaged `rrsync` script
* Make some improvements/fixes following `shellcheck` and `hadolint`
* Add a build script for maintenance
* Document usage and maintenance

See commit messages for details.

Also updated [image tags on Docker Hub](https://hub.docker.com/r/metabrainz/sshd/tags):
* Added the tag `v-2016-11-08` to the previous image (tagged `latest` so far)
* Built from this branch and pushed under the tag `v-2025-02-24.0`

## Test

Locally ran the new image as follows:
```bash
docker run \
    --env SSHD_LISTEN_ADDRESS=127.0.0.1 \
    --env SSHD_PORT=22345 \
    --name test-sshd-jammy \
    --rm \
    metabrainz/sshd:v-2025-02-24.0
```

And checked from a separate shell with:
```bash
docker exec -it test-sshd-jammy apt list --installed openssh-server rsync
# openssh-server/now 1:8.9p1-3ubuntu0.11 amd64 [installed,local]
# rsync/now 3.2.7-0ubuntu0.22.04.4 amd64 [installed,local]

docker exec -it test-sshd-jammy bash -c 'ps -p $(pidof sshd) -o args'
# COMMAND
# sshd: /usr/sbin/sshd -D -o ListenAddress=127.0.0.1 -o Port=22345 [listener] 0 of 10-100 startups

```

## Further actions

* [ ] Set image tag in deployment

yvanzo committed:
Drop redundant/outdated copy of the script `rrsync`

The script `rrsync` was a copy of an upstream script back in the time.
An updated version of this script is provided by the package `rsync`
for years (roughly one week after the latest build of this repository).
Thus it is no longer helpful to copy it, quite the contrary.

yvanzo committed:
Pin base image tag (currently based on Jammy)

Specify the tag used for the base image (currently based on Jammy),
thus following best practices, as the tag `latest` is not reliable.

References:
* https://github.com/hadolint/hadolint/wiki/DL3006
* https://docs.docker.com/build/building/best-practices/#pin-base-image-versions

yvanzo committed:
Lighten installing packages with apt-get

- Avoid additional packages by specifying `--no-install-recommends`
- Delete the apt-get lists after installing something

References:
* https://github.com/hadolint/hadolint/wiki/DL3009
* https://github.com/hadolint/hadolint/wiki/DL3015
* https://docs.docker.com/build/building/best-practices/#apt-get

yvanzo committed:
Fail build even on error from piped command

yvanzo committed:
Add script to build and push image to Docker hub

This script is based on the script from our docker-irccat repository,
with changes adviced by `shellcheck`.

The default date versioning follows our current common practices (in
listenbrainz-server, musicbrainz-server…) for tagging Docker images.

yvanzo committed:
Always build with the latest security updates

Intentionally builds image with no cache (but pulled base-image),
in order to upgrade packages with the latest security updates.

yvanzo committed:
Ignore irrelevant warning about pining packages

The alternative approach recommended by `hadolint` is incompatible
with upgrading packages which is also among Docker’s best practices.

References:
* https://github.com/hadolint/hadolint/wiki/DL3008
* https://docs.docker.com/build/building/best-practices/#apt-get

yvanzo committed:
Document usage and maintenance

zas commented on push.sh:
```diff
@@ -0,0 +1,35 @@
+${DOCKER_CMD} push "${image_name}:${image_tag}" \
+  2>&1 | tee -a ./"build-${image_tag}-at-${timestamp}.log"
+
+# vi: set et sts=2 sw=2 ts=2 :
```
While this works, what about using a github action instead?
Something like: https://github.com/metabrainz/picard-website/blob/master/.github/workflows/docker-ci.yml

Basically you just push a tag and the action build and push to docker hub.
For picard-website version tags starts with v[0-9] (match is done in https://github.com/metabrainz/picard-website/blob/be7456fdef7053639a6946eb3ed1445471cdf6d2/.github/workflows/docker-ci.yml#L7) but that can be easily modified.

yvanzo replied:
Does it help with building new images even when no change occurred to the repository?
That would be great. I never used it before so it is going to take some time. Probably not worth waiting for it.

zas replied:
> Does it help with building new images even when no change occurred to the repository?

If you want to force building just push another tag (pointing to same commit). We often do that for Picard Website: even if there's no change in website code, the image build process updates plugins.

yvanzo replied:
Just added https://github.com/metabrainz/docker-sshd/pull/1/commits/bf5dd4a76a4a2613c882c24751cb57de47d30b7f

Tested the tagging script only.

yvanzo committed:
Publish to Docker Hub via GitHub Actions workflow

Compared to picard-website’s workflow, the differences are:
* Cache is disabled (to get the latest security updates)
* No other trigger than git tags

It replaces the script `push.sh`.
It comes with a new script `tag.sh` for brainless tagging.

zas reviewed (approved):
LGTM
